### PR TITLE
[Monit-28482] Exponential histogram

### PR DESCRIPTION
**Pull request by deosu**
- Code changes for both delta & cumulative exponential histogram
- Code coverage

Travis & Sumit

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 1]
```diff
@@ -195,41 +203,43 @@ private static List<ReportPoint> transformHistogram(String name, Histogram histo
     }
   }
 
-  private static List<ReportPoint> transformDeltaHistogram(String name, Histogram histogram, List<KeyValue> resourceAttrs) {
+  private static List<ReportPoint> transformDeltaHistogram(String name, IHistogram histogram, List<KeyValue> resourceAttrs) {
     List<ReportPoint> reportPoints = new ArrayList<>();
-    for (HistogramDataPoint point : histogram.getDataPointsList()) {
-      reportPoints.addAll(transformDeltaHistogramDataPoint(name, point, resourceAttrs));
+    for (int i = 0; i < histogram.numDataPoints(); i++) {
+      reportPoints.addAll(transformDeltaHistogramDataPoint(name, histogram.at(i), resourceAttrs));
```
[the remark this answers is not in the document]
Comment (keep94) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:209]: Good question, thepeterstone. We used the same ideas in the OTEL collector here. It is essentially a direct port from the GO code.  To use the foreach-style loop we would need some method that returns an Iterable<IHistogramDataPoint>. We would need to decide if our Iterable would be eager, e.g an ArrayList<IHistogramDataPoint> where we make all the IHistogramDataPoint interfaces all at once, or lazy where we make the IHistogramDataPoint interfaces as we iterate.  Doing it the way we did seemed easier, but perhaps less JAVA like.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 2]
```diff
@@ -457,12 +409,61 @@ private static ReportPoint.Builder pointWithAnnotations(String name, List<KeyVal
     return builder;
   }
 
+  /**
+   * IHistogram is a common interface that both regular histograms and exponential histograms can be
+   * converted to. This allows us to write code that works for both regular and exponential histograms.
+   */
+  private interface IHistogram {
```
[the remark this answers is not in the document]
Comment (keep94) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:416]: As we discussed, we now have just one new class: BucketHistogramDataPoint.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 3]
```diff
@@ -44,6 +47,7 @@
 
 
 public class OtlpProtobufPointUtils {
+
```
[the remark this answers is not in the document]
Comment (keep94) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:50]: Deleted. Its disturbing that command-option-shift-L doesn't fix errors like this.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (keep94) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:209]: With our recent re-factoring, I brought back the for-each loop.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 4]
```diff
@@ -195,41 +203,43 @@ private static List<ReportPoint> transformHistogram(String name, Histogram histo
...
   private static List<ReportPoint> transformDeltaHistogramDataPoint(String name,
-                                                                    HistogramDataPoint point, List<KeyValue> resourceAttrs) {
-    if (point.getExplicitBoundsCount() != point.getBucketCountsCount() - 1) {
+                                                                    IHistogramDataPoint point, List<KeyValue> resourceAttrs) {
+    List<Double> explicitBounds = point.getExplicitBounds();
+    List<Long> bucketCounts = point.getBucketCounts();
+    if (explicitBounds.size() != bucketCounts.size() - 1) {
       throw new IllegalArgumentException("OTel: histogram " + name + ": Explicit bounds count " +
-          "should be one less than bucket count. ExplicitBounds: " + point.getExplicitBoundsCount() +
-          ", BucketCounts: " + point.getExplicitBoundsCount());
```
[the remark this answers is not in the document]
Comment (keep94) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:222]: Yes. Thank you.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 5]
```diff
@@ -397,4 +408,249 @@ private static ReportPoint.Builder pointWithAnnotations(String name, List<KeyVal
...
+     */
+    List<KeyValue> getAttributesList();
+
+    /**
+     * @return the timestamp for this data point.
+     */
+    long getTimeUnixNano();
+  }
+
+  private static class RegularHistogram implements IHistogram {
```
[the remark this answers is not in the document]
Comment (keep94) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:460]: We now just have one new class: BucketHistogramDataPoint. No more class hierarchy.

**Comment by keep94** [comment]
thepeterstone, thanks for your feedback. Because of your ideas, the code is now 83 lines shorter.

**Review by thepeterstone — APPROVED** [review]
LGTM

**Comment by keep94** [comment]
Superseded by PR #18

**Pull request closed without merging**